File: projects/01-url-shortener/src/metrics.rs
//! Prometheus metrics for the observability checklist.
//!
//! The [`metrics`] facade writes to a process-global recorder, so the counter/
//! gauge call sites (in [`routes`](crate::routes) and [`ingest`](crate::ingest))
//! stay decoupled from this wiring — they just name a metric. Until [`install`]
//! sets a recorder the macros are no-ops, which is exactly what tests want (no
//! metrics setup, no panics).
//!
//! [`install`] returns a [`PrometheusHandle`]; render it from a `/metrics` route
//! (see [`crate::routes::metrics_router`]) for a scrape endpoint. The three
//! metrics the SPEC grades:
//! - [`REDIRECTS_TOTAL`] — redirects served, labelled by cache outcome.
//! - [`CACHE_LOOKUPS_TOTAL`] — every redirect resolution, labelled by outcome;
//!   `hit / sum` is the cache hit ratio.
//! - [`INGEST_QUEUE_DEPTH`] — live depth of the click-ingestion channel.

use metrics_exporter_prometheus::{PrometheusBuilder, PrometheusHandle};

/// Redirects served (a 3xx was returned), labelled `cache = hit|miss`.
pub const REDIRECTS_TOTAL: &str = "url_shortener_redirects_total";

/// Cache-aside resolutions on the redirect path, labelled `outcome =
/// hit|miss|negative`. The hit ratio is `hit / (hit + miss + negative)`.
pub const CACHE_LOOKUPS_TOTAL: &str = "url_shortener_cache_lookups_total";

/// Number of buffered clicks waiting in the ingestion channel (a gauge:
/// incremented on accept, decremented as the ingestor drains).
pub const INGEST_QUEUE_DEPTH: &str = "url_shortener_ingest_queue_depth";

/// Install the process-global Prometheus recorder and return a handle used to
/// render the registry for `/metrics`. Call once, from `main`, after telemetry
/// init. Panics if a recorder is already installed (calling it twice is a bug).
pub fn install() -> PrometheusHandle {
    register_descriptions();
    PrometheusBuilder::new()
        .install_recorder()
        .expect("install prometheus recorder")
}

/// Register HELP metadata for the three SPEC metrics. Call sites only pass the
/// name constant; describing them here keeps the strings single-sourced and
/// gives rustc a direct use of each constant (proc-macro call sites alone do
/// not always satisfy `dead_code`).
fn register_descriptions() {
    let redirects = REDIRECTS_TOTAL;
    let cache_lookups = CACHE_LOOKUPS_TOTAL;
    let ingest_queue = INGEST_QUEUE_DEPTH;
    metrics::describe_counter!(
        redirects,
        "Redirects served (3xx returned), labelled by cache outcome"
    );
    metrics::describe_counter!(
        cache_lookups,
        "Cache-aside resolutions on the redirect path, labelled by outcome"
    );
    metrics::describe_gauge!(
        ingest_queue,
        "Buffered clicks waiting in the ingestion channel"
    );
}

File: projects/01-url-shortener/src/url_shortener/metrics.py
"""Prometheus metrics for the observability checklist.

`prometheus_client` collectors register themselves into the default registry at
import time, and `common_telemetry.metrics_routes()` renders that same registry
at `/metrics` - so declaring a metric here is all the wiring there is. The three
the SPEC grades:

* :data:`REDIRECTS_TOTAL` - redirects served, labelled by cache outcome.
* :data:`CACHE_LOOKUPS_TOTAL` - every resolution, labelled by outcome; `hit`
  over the sum is the cache hit ratio.
* :data:`INGEST_QUEUE_DEPTH` - live depth of the click-ingestion queue.

Note the shape of the first two: a *counter per outcome*, not a pre-computed
ratio. A ratio cannot be aggregated across replicas or re-windowed after the
fact; two counters can, and Prometheus does the division at query time.
"""

from __future__ import annotations

from prometheus_client import Counter, Gauge

__all__ = ["CACHE_LOOKUPS_TOTAL", "INGEST_QUEUE_DEPTH", "REDIRECTS_TOTAL"]

REDIRECTS_TOTAL = Counter(
    "url_shortener_redirects_total",
    "Redirects served (3xx returned), labelled by cache outcome",
    ["cache"],
)

CACHE_LOOKUPS_TOTAL = Counter(
    "url_shortener_cache_lookups_total",
    "Cache-aside resolutions on the redirect path, labelled by outcome",
    ["outcome"],
)

INGEST_QUEUE_DEPTH = Gauge(
    "url_shortener_ingest_queue_depth",
    "Buffered clicks waiting in the ingestion queue",
)
